ni_measurementlink_service/_drivers/__init__.py

```diff
@@ -16,9 +20,12 @@ def closing_session(session: TSession) -> ContextManager[TSession]:
     Returns:
         A context manager that yields the session and closes it.
     """
-    if isinstance(session, contextlib.AbstractContextManager):
+    if isinstance(session, contextlib.AbstractContextManager) and initialization_behavior == SessionInitializationBehavior.AUTO:
```

Comment:
As described in this HLD, the ideal behavior for AUTO is "to Init then Close if the instrument session is not existed on grpc-device server, attach then detach if the instrument session is existed on grpc-device server" but the current implementation doesn't fully incorporate this behavior.

I'll continue on exploring that part and will update the PR in the coming days. The other 4 enum behaviour is showcased for the niscope session with TestStand code module changes.

Reply:
I might be missing something, but it seems like what you have here should more or less work. I don't know if this is also used by third party drivers or just NI drivers. 

My only other comment is this code makes certain assumptions about the driver session object's behavior. For instance, the context manager implemented by the session for VISA could likely be used as is all of the time. Similarly, I could see how some drivers might have a `Detach` API that should be called rather than just not calling anything. If you wanted to allow for that, I think you would have to add adapter code in each driver's `SessionConstructor` class rather than trying to share the code. I guess you could still call a shared function from the driver specific constructor if there are common implementation patterns used by various drivers.